waftools/detections/compiler_swift.py

```diff
@@ -19,50 +21,162 @@ def __add_swift_flags(ctx):
     verRe = re.compile("(?i)version\s?([\d.]+)")
     ctx.env.SWIFT_VERSION = verRe.search(__run([ctx.env.SWIFT, '-version'])).group(1)
 
-    # the -swift-version parameter is only supported on swift 3.1 and newer
-    if StrictVersion(ctx.env.SWIFT_VERSION) >= StrictVersion("3.1"):
-        ctx.env.SWIFT_FLAGS.extend([ "-swift-version", "3" ])
+    # prevent possible breakages with future swift versions
+    if StrictVersion(ctx.env.SWIFT_VERSION) >= StrictVersion("6.0"):
+        ctx.env.SWIFT_FLAGS.extend([ "-swift-version", "5" ])
 
     if ctx.is_debug_build():
         ctx.env.SWIFT_FLAGS.append("-g")
 
     if ctx.is_optimization():
         ctx.env.SWIFT_FLAGS.append("-O")
 
-def __add_swift_library_linking_flags(ctx, swift_library):
+def __add_static_swift_library_linking_flags(ctx, swift_library):
     ctx.env.append_value('LINKFLAGS', [
         '-L%s' % swift_library,
         '-Xlinker', '-force_load_swift_libs', '-lc++',
     ])
 
+def __add_dynamic_swift_library_linking_flags(ctx, swift_library):
+    ctx.env.append_value('LINKFLAGS', [ '-L%s' % swift_library ])
+
+    #ABI compatibility
+    if StrictVersion(ctx.env.SWIFT_VERSION) >= StrictVersion("5.0"):
+        ctx.env.append_value('LINKFLAGS', [
+            '-Xlinker', '-rpath', '-Xlinker', '/usr/lib/swift',
+        ])
+
+    ctx.env.append_value('LINKFLAGS', [
+        '-Xlinker', '-rpath', '-Xlinker', swift_library,
+    ])
+
 def __find_swift_library(ctx):
-    swift_library_paths = [
-        'Toolchains/XcodeDefault.xctoolchain/usr/lib/swift_static/macosx',
-        'usr/lib/swift_static/macosx'
-    ]
+    swift_libraries = {}
+    #look for set lib paths in passed environment variables
+    if 'SWIFT_LIB_DYNAMIC' in ctx.environ:
+        swift_libraries['SWIFT_LIB_DYNAMIC'] = ctx.environ['SWIFT_LIB_DYNAMIC']
+    if 'SWIFT_LIB_STATIC' in ctx.environ:
+        swift_libraries['SWIFT_LIB_STATIC'] = ctx.environ['SWIFT_LIB_STATIC']
+
+    #search for swift libs relative to the swift compiler executable
+    swift_library_relative_paths = {
+        'SWIFT_LIB_DYNAMIC': '../../lib/swift/macosx',
+        'SWIFT_LIB_STATIC': '../../lib/swift_static/macosx'
+    }
+
+    for lib_type, path in swift_library_relative_paths.items():
+        if lib_type not in swift_libraries:
+            lib_path = os.path.join(ctx.env.SWIFT, path)
+            swift_library = ctx.root.find_dir(lib_path)
+            if swift_library is not None:
+                swift_libraries[lib_type] = swift_library.abspath()
+
+    #fall back to xcode-select path
+    swift_library_paths = {
+        'SWIFT_LIB_DYNAMIC': [
```

Comment:
This type of dict with keys containing a list of strings instead of just a string could be normalized upon, in which case the looping over them could be put into a function or so and the code could be re-used instead of two similar-yet-different pieces of code existing.